modin/core/dataframe/base/dataframe/dataframe.py

```diff
@@ -0,0 +1,506 @@
+        axis : int
+            The axis to filter over.
+        condition : callable
+            The function to use for the filter. This function should filter the
+            data itself.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe from the filter provided.
+        """
+        pass
+
+    @abstractmethod
+    def explode(
+        self,
+        axis: int,
+        function: Callable,
+        result_schema: Optional[Dict[Hashable, type]] = None,
+    ) -> "ModinDataframe":
+        """
+        Explode data based on the function provided along the specified axis.
+
+        Parameters
+        ----------
+        axis : int
+            The axis to expand over.
+        function : callable
+            The function to use to expand the data. This function should accept one
+            row/column, and return multiple.
+        result_schema : dictionary, default: None
+            Mapping from column labels to data types that represents the types of the output dataframe.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe with the specified axis expanded.
+
+        Notes
+        -----
+        Only one axis can be expanded at a time.
+
+        The user-defined function may increase the number of rows (columns if axis=1),
+                but it should not remove or drop rows.
+        """
+        pass
+
+    @abstractmethod
+    def window(
+        self,
+        axis: int,
+        reduce_fn: Callable,
+        window_size: int,
+        result_schema: Optional[Dict[Hashable, type]] = None,
+    ) -> "ModinDataframe":
+        """
+        Apply a sliding window operator that acts as a GROUPBY on each window, reducing each window to a single row (column).
+
+        Parameters
+        ----------
+        axis : int
+            The axis to slide over.
+        reduce_fn : callable
+            The reduce function to apply over the data.
+        window_size : int
+            The number of row/columns to pass to the function.
+            (The size of the sliding window).
+        result_schema : dictionary, default: None
+            Mapping from column labels to data types that represents the types of the output dataframe.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe with the reduce function applied over windows of the specified
+                axis.
+
+        Notes
+        -----
+        The user-defined reduce function must reduce each window’s column
+            (row if axis=1) down to a single value.
+        """
+        pass
+
+    @abstractmethod
+    def groupby(
+        self,
+        axis: int,
+        by: Union[str, List[str]],
+        operator: Callable,
+        result_schema: Optional[Dict[Hashable, type]] = None,
+    ) -> "ModinDataframe":
+        """
+        Generate groups based on values in the input column(s) and perform the specified operation on each.
```

Comment:
@devin-petersohn perhaps we should add a level parameter? That's how pandas gets around this

Reply:
`by` as a level should not be supported here explicitly. From the API layer, it would be done via the following algebraic operators:

1.) `from_labels`
2.) `groupby(by=level_name)`
3.) `to_labels` (if needed)

The goal here is to simplify the way that the labels are created for groupby. The logic as-is is too complicated and very messy. cc @dchigarev